spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java

```diff
@@ -464,6 +466,54 @@ public void shouldNotImplementedWithRegOnlyOfferTypesAndNoOwnerForV2Auth(String
             .cloudAuthorizeV2("account-id", "cloud-instance-id", offerId, "test-type", ""));
     }
 
+    @Test
+    public void shouldAllowCloudRegistrationWithNoOfferingIds() throws Exception {
+        ApiClient adminClient = ApiClients.admin();
+
+        String accountId = StringUtil.random("cloud-account-id-");
+        String instanceId = StringUtil.random("cloud-instance-id-");
+        String offerId = StringUtil.random("cloud-offer-");
+
+        OwnerDTO owner = adminClient.owners().createOwner(Owners.random());
+        ProductDTO prod = adminClient.ownerProducts().createProduct(owner.getKey(), Products.random());
+        adminClient.owners().createPool(owner.getKey(), Pools.random(prod));
+
+        adminClient.hosted().createOwner(owner);
+        adminClient.hosted().createProduct(prod);
+        adminClient.hosted().createSubscription(Subscriptions.random(owner, prod));
+        adminClient.hosted().associateOwnerToCloudAccount(accountId, owner.getKey());
+        adminClient.hosted().associateProductIdsToCloudOffer(offerId, List.of(prod.getId()));
+
+        CloudAuthenticationResultDTO cloudAuthDTO = ApiClients.noAuth().cloudAuthorization()
+            .cloudAuthorizeV2(accountId, instanceId, offerId, "test-type", "");
```

Comment:
As I mentioned elsewhere, this issue is unrelated to autoregistration. It is related to registration of a cloud VM (which will send cloud-related facts), regardless of the type of auth (you could move this test to `ConsumerResourceSpecTest` and use username/password or AK; as long as the consumer has those aws fact keys with null values, you'll be able to reproduce)

Reply:
Ugh...

So do we care about the cloud registration details when it's not a cloud registration request? Seems like what you're describing is a second bug where during a non-cloud registration, we're still treating it like one and processing facts very specific to cloud registration.

Additionally, because we opted to have very highly specific facts for each cloud provider, we probably need tests for every set of cloud facts. Worse, I fear that combinations of them are also "valid".

Reply:
I've explained in more detail in the relevant slack thread, but the cloud facts _are not used/related_ to autoregistration. They're used for reporting reasons by other upstream systems (e.g. Swatch), for any cloud system that registers, regardless if they used autoreg or other means of registration

Reply:
Please link the slack thread in the Jira so it can be tracked.

That aside, we need to be a lot more clear on our terminology. Facts are generally not operative parameters for us, and calling a system a "cloud system" heavily implies cloud registration -- especially given everything else around this.

Now, rewinding a bit, if these are only used for reporting and are non-operative, why are we putting so much extra, multi-class processing behind these? Why are they stored separately? What happens if the facts update later? What if the values desync from an actual cloud registration (where appropriate)? What if more are added later?

There are so many questions given what you're implying about their lack of importance relative to how much code and state exists surrounding them.

Reply:
- They're stored separately because it is impossible to do any performant querying against the facts table.
- The current multi-class processing (especially the excessive validation) is a miss on the initial implementation/review.
- These facts never change, and this was taken into account when designing this (no need for updates to these values). A cloud VM will always belong to the same cloud provider, cloud account, have the same cloud instance id, and the same cloud offering ids. You can't 'move' cloud vms from AWS to Azure for example, and you can't move it from one AWS account to another. The cloud instance id can't change because it is the unique id assigned for it. The cloud offering ids won't change because they are linked to a specific cloud image type, if that changed, that would mean it is an entirely different VM based on an entirely different cloud image type.

Reply:
For the first point: true enough, I suppose. But while our storage issues are something we aim to address in the future, any convention or specification we defined now is still something we have to carry forward. And if we get to a place where we're back to querying against facts directly instead of storing facts we care about specially, we're back to this weird point of some facts being special and having special logic and behavior, and some not.

For the last point: it doesn't matter what the client conventions are (today) when *our* limitations are very much not that. Facts can change arbitrarily, and it's little more than a key/value pair from our perspective. Assuming the client will never change nor misbehave is a grave error in our design.

Glancing through the code it looks like we pull the facts once during registration (and in a weird place relative to all the other input processing), and then never again. So while the facts could change, we won't ever process them again as cloud facts; which itself is really weird.

Reply:
It wasn't really client conventions, but that cloud provider metadata services wouldn't be updating this data for a specific cloud VM logically that was used to decide not to have these being updatable. But I guess adding logic to make them updateable in candlepin as a guard against subman bugs is a good idea (maybe as a separate task than this bug fix?)

Reply:
The scope of this PR is to fix the immediate blocker (registration failing for all cloud systems against Stage CP), when cloud facts which have a key but a null value are sent

Reply:
I've added tests for all sets of cloud facts for both cloud registration and standard registration.